### DISCUSSION TITLE
Problem with fanflat geometry

### DISCUSSION THREAD

**Discussion by maikeros** [astra-toolbox/astra-toolbox#411] [Q&A]
Hello there, 

I simulated a CT with vGate and am using .txt files with pixel inetsities (matrix) as the input for image reconstruction. I got a code for cone beam reconstruction to work. 

from __future__ import division
from imageio import imread, imwrite
import cv2
import numpy as np
from os import mkdir
from os.path import join, isdir
from imageio import get_writer
import imageio.v2 as imageio
import astra
import os 

# Konfiguration
distance_source_origin = 150  # [mm]
distance_origin_detector = 150  # [mm]
detector_pixel_size = 0.5  # [mm]
detector_rows = 100  # Vertikale Größe des Detektors [Pixel].
detector_cols = 100  # Horizontale Größe des Detektors [Pixel].
num_of_projections = 30
angles = np.linspace(0, 2 * np.pi, num=num_of_projections, endpoint=False)
input_dir = 'txt_files'  # Ordner, in dem Ihre Textdateien liegen
output_dir = 'reconstruction'  # Ordner für die Rekonstruktion

# Laden der Projektionen
projections = np.zeros((detector_rows, num_of_projections, detector_cols))
for i in range(num_of_projections):
    # Annahme: Ihre Projektionsdaten sind in Textdateien mit dem Namen 'proj%04d.txt' gespeichert
    # Verwenden Sie die richtigen Dateinamen und Pfade entsprechend Ihrer Dateistruktur
    data = np.loadtxt(join(input_dir, 'runID%d.txt' % i))
    projections[:, i, :] = data

# Kopieren der Projektionsbilder in die ASTRA Toolbox
proj_geom = \
  astra.create_proj_geom('cone', 1, 1, detector_rows, detector_cols, angles,
                         (distance_source_origin + distance_origin_detector) /
                         detector_pixel_size, 0)
projections_id = astra.data3d.create('-sino', proj_geom, projections)

# Erstellen der Rekonstruktion
vol_geom = astra.creators.create_vol_geom(detector_cols, detector_cols,
                                          detector_rows)
reconstruction_id = astra.data3d.create('-vol', vol_geom, data=0)
alg_cfg = astra.astra_dict('FDK_CUDA')
alg_cfg['ProjectionDataId'] = projections_id
alg_cfg['ReconstructionDataId'] = reconstruction_id
algorithm_id = astra.algorithm.create(alg_cfg)
astra.algorithm.run(algorithm_id)
reconstruction = astra.data3d.get(reconstruction_id)

# Beschränken und skalieren der Rekonstruktion
reconstruction[reconstruction < 0] = 0
reconstruction /= np.max(reconstruction)
reconstruction = np.round(reconstruction * 255).astype(np.uint8)

# Rekonstruktion speichern
if not isdir(output_dir):
    mkdir(output_dir)
for i in range(detector_rows):
    im = reconstruction[i, :, :]
    im = np.flipud(im)
    imwrite(join(output_dir, 'reco%04d.png' % i), im)

# Aufräumen
astra.algorithm.delete(algorithm_id)
astra.data3d.delete(reconstruction_id)
astra.data3d.delete(projections_id)

# Schleife, um die rekonstruierten Bilder anzuzeigen
for i in range(detector_rows):
    im = reconstruction[i, :, :]
    im = np.flipud(im)
    cv2.imshow('Reconstruction', im)
    cv2.waitKey(100)  # Warten Sie für 100 Millisekunden, bevor Sie das nächste Bild anzeigen

cv2.destroyAllWindows()  # Schließen Sie das Fenster, wenn Sie fertig sind

Now I tried to change it to fanflat geometry reconstruction, because I did a fanbeam CT and I only get error responses.

from __future__ import division
from imageio import imwrite
import cv2
import numpy as np
from os import mkdir
from os.path import join, isdir
import astra

# Konfiguration
distance_source_origin = 150  # [mm]
distance_origin_detector = 150  # [mm]
detector_pixel_size = 0.5  # [mm]
detector_rows = 100  # Vertikale Größe des Detektors [Pixel].
detector_cols = 100  # Horizontale Größe des Detektors [Pixel].
num_of_projections = 30
angles = np.linspace(0, 2 * np.pi, num=num_of_projections, endpoint=False)
input_dir = 'txt_files'  # Ordner, in dem Ihre Textdateien liegen
output_dir = 'reconstruction'  # Ordner für die Rekonstruktion

# Laden der Projektionen
projections = np.zeros((detector_rows, num_of_projections, detector_cols))
for i in range(num_of_projections):
    # Annahme: Ihre Projektionsdaten sind in Textdateien mit dem Namen 'proj%04d.txt' gespeichert
    # Verwenden Sie die richtigen Dateinamen und Pfade entsprechend Ihrer Dateistruktur
    data = np.loadtxt(join(input_dir, 'runID%d.txt' % i))
    projections[:, i, :] = data

# Kopieren der Projektionsbilder in die ASTRA Toolbox
proj_geom = astra.create_proj_geom('fanflat', detector_pixel_size, detector_pixel_size, detector_rows, detector_cols, angles, distance_source_origin, distance_origin_detector)

projections_id = astra.data3d.create('-sino', proj_geom, projections)

# Erstellen der Rekonstruktion
vol_geom = astra.create_vol_geom(detector_cols, detector_cols, detector_rows)
reconstruction_id = astra.data3d.create('-vol', vol_geom, data=0)
alg_cfg = astra.astra_dict('FBP_CUDA')
alg_cfg['ProjectionDataId'] = projections_id
alg_cfg['ReconstructionDataId'] = reconstruction_id
algorithm_id = astra.algorithm.create(alg_cfg)
astra.algorithm.run(algorithm_id)
reconstruction = astra.data3d.get(reconstruction_id)

# Beschränken und skalieren der Rekonstruktion
reconstruction[reconstruction < 0] = 0
reconstruction /= np.max(reconstruction)
reconstruction = np.round(reconstruction * 255).astype(np.uint8)

# Rekonstruktion speichern
if not isdir(output_dir):
    mkdir(output_dir)
for i in range(detector_rows):
    im = reconstruction[i, :, :]
    im = np.flipud(im)
    imwrite(join(output_dir, 'reco%04d.png' % i), im)

# Aufräumen
astra.algorithm.delete(algorithm_id)
astra.data3d.delete(reconstruction_id)
astra.data3d.delete(projections_id)

# Schleife, um die rekonstruierten Bilder anzuzeigen
for i in range(detector_rows):
    im = reconstruction[i, :, :]
    im = np.flipud(im)
    cv2.imshow('Reconstruction', im)
    cv2.waitKey(100)  # Warten Sie für 100 Millisekunden, bevor Sie das nächste Bild anzeigen

cv2.destroyAllWindows()  # Schließen Sie das Fenster, wenn Sie fertig sind

python test.py
Traceback (most recent call last):
  File "/home/student/Desktop/Maike/astra/test.py", line 31, in <module>
    projections_id = astra.data3d.create('-sino', proj_geom, projections)
  File "/home/student/anaconda3/lib/python3.10/site-packages/astra/data3d.py", line 43, in create
    return d.create(datatype,geometry,data)
  File "astra/data3d_c.pyx", line 115, in astra.data3d_c.create
ValueError: Invalid geometry type.

Does anyone has an idea, how to solve it? I use astra toolbox version 2.1.2.

Kind regards,
Maike

**Comment by askorikov** [comment]
Hi! I cannot reproduce the issue, seems to work for me.

**Comment by maikeros** [comment]
Thank you, for your quick answer!
And you really used the fan beam code with 30 .txt files with each 100x100 entries in it and it worked? 
That' really weird, because I still get the error message.

**Comment by askorikov** [comment]
Ah no, sorry. I was checking the first part of the script only (I missed the transition in the middle). In the second part, the error is thrown because `fanflat` in Astra is a 2D geometry. For 3D data, you can apply `FBP_CUDA` algorithm to each reconstruction slice in a loop while using 2D data objects.

**Comment by askorikov** [comment]
I've also tried to improve the error message for this case in [6d09148](https://github.com/astra-toolbox/astra-toolbox/pull/397/commits/6d09148b9eed1b5475ab1b22c84706dd687b2ec5).
